kubelet: fix sandbox garbage collection

feiskyer opened the pull request:
**What this PR does / why we need it**:

Currently, kuberuntime garbage collection can't distinguish just-created sandboxes from failed sandboxes. Especially when the time from sandbox creation to ready is longer than GC's minAge. In such cases, those sandboxes may be garbage collected early before they are ready.

This PR removes `sandboxMinGCAge` and only garbage collect sandboxes when 

* they are containing no containers at all
* and not the latest sandbox if it is belonging to an existing pod.

**Which issue this PR fixes** 

Fixes #42856.

**Release note**:

```release-note
NONE
```

cc @yujuhong @Random-Liu

feiskyer committed:
Fix sandbox garbage collection.

Sandboxes are garbage collected only when they are containing no containers at
all and not the latest sandbox if it is belonging to an existing pod.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43053)
<!-- Reviewable:end -->

feiskyer commented:
/assign @yujuhong @Random-Liu

feiskyer commented:
@k8s-bot kops aws e2e test this

feiskyer commented:
@k8s-bot kops aws e2e test this

feiskyer commented:
ping @yujuhong @Random-Liu

feiskyer commented:
ping @Random-Liu @yujuhong

Random-Liu commented:
Will take a look tomorrow.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -128,6 +130,18 @@ func (cgc *containerGC) removeOldestN(containers []containerGCInfo, toRemove int
+	// Remove from oldest to newest (last to first).
+	numToKeep := len(sandboxes) - toRemove
+	for i := numToKeep; i < len(sandboxes); i++ {
+		cgc.removeSandbox(sandboxes[i].id)
```
Why don't we log error here?

feiskyer replied:
There is already such log inside `cgc.removeSandbox`

Random-Liu replied:
I see.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -252,39 +267,43 @@ func (cgc *containerGC) evictSandboxes(minAge time.Duration) error {
+			// Remove all sandboxes if pod has been deleted.
+			cgc.removeOldestNSandboxes(sandboxes, len(sandboxes))
+		} else {
+			// Keep latest one if pod is still exist.
```
s/is still exist/still exists

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -252,39 +267,43 @@ func (cgc *containerGC) evictSandboxes(minAge time.Duration) error {
+			cgc.removeOldestNSandboxes(sandboxes, len(sandboxes)-1)
+		}
 	}
+
```
nit: Unnecessary empty line.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -52,55 +68,49 @@ func TestSandboxGC(t *testing.T) {
 			sandboxes: []sandboxTemplate{
-				{pod: pods[0], state: runtimeapi.PodSandboxState_SANDBOX_NOTREADY},
+				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY),
+				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_READY),
```
Not sure whether we should remove READY sandbox. `SyncLoop` may be stopping the sandbox, too. This may introduce race that sandbox is stopped twice.

feiskyer replied:
Yes, didn't notice this problem. Will update to omit the ready sandbox.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -252,39 +267,43 @@ func (cgc *containerGC) evictSandboxes(minAge time.Duration) error {
-	evictSandboxes := make([]string, 0)
-	newestGCTime := time.Now().Add(-minAge)
+	evictSandboxes := make(sandboxesByPodUID)
 	for _, sandbox := range sandboxes {
```
Why not skip running sandbox?

feiskyer replied:
The new logic only keeps the latest sandbox for each pod, all other sandboxes (including running ones) should be removed.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -34,6 +34,19 @@ func TestSandboxGC(t *testing.T) {
 
+	fakePodGetter := m.containerGC.podGetter.(*fakePodGetter)
+	makeGCSandbox := func(pod *v1.Pod, attempt uint32, state runtimeapi.PodSandboxState) sandboxTemplate {
+		if pod.Name != "deleted" {
```
Why not introduce an arguement to indicate whether to add the pod into pod getter?

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -52,55 +68,49 @@ func TestSandboxGC(t *testing.T) {
 			sandboxes: []sandboxTemplate{
 				// running sandbox.
-				{pod: pods[0], attempt: 1, state: runtimeapi.PodSandboxState_SANDBOX_READY},
+				makeGCSandbox(pods[0], 1, runtimeapi.PodSandboxState_SANDBOX_READY),
```
I don't understand, why the `2` is not deleted here? We only keep the latest sandbox right?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -239,9 +253,10 @@ func (cgc *containerGC) evictContainers(gcPolicy kubecontainer.ContainerGCPolicy
-// and contains no containers at all.
-func (cgc *containerGC) evictSandboxes(minAge time.Duration) error {
+// evictSandboxes evicts all sandboxes that are evictable. Evictable sandboxes are:
+// containing no containers at all and not the latest sandbox if it is belonging to
```
Reword a little bit to make this more clear:
```
// evictSandboxes remove all evictable sandboes. An evictable sandbox must
// meet the following requirements:
//   1. contains no containers.
//   2. belong to a non-existent (i.e., already removed) pod, or is not the
//      most recently created sandbox for the pod.
```

yujuhong commented:
It's a shame because I actually like the original implementation which removes the podsandbox only when there are no associated containers in it. I think that'd be the right model as we *slowly* move away from using dead container bodies as tombstones.

feiskyer commented:
@Random-Liu @yujuhong Addressed comments. PTAL.

> It's a shame because I actually like the original implementation which removes the podsandbox only when there are no associated containers in it. I think that'd be the right model as we slowly move away from using dead container bodies as tombstones.

@yujuhong That's also my initial proposal,  but it needs a new sandbox state (and of course new sandbox APIs) to do that way, see [here](https://github.com/kubernetes/kubernetes/issues/42856#issuecomment-286012484). I think keeping  latest pod sandbox (in this PR) is better because it follows same logic as we do for container GC.

feiskyer commented:
@k8s-bot cvm gce e2e test this

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -253,9 +253,12 @@ func (cgc *containerGC) evictContainers(gcPolicy kubecontainer.ContainerGCPolicy
-// evictSandboxes evicts all sandboxes that are evictable. Evictable sandboxes are:
-// containing no containers at all and not the latest sandbox if it is belonging to
-// an existing pod.
+// evictSandboxes remove all evictable sandboes. An evictable sandbox must
```
I believe this function is wrong.

For container garbage collection, what we are doing:
1) get `evictableContainers` (prune out running ones);
2) remove all or keep latest `evictableContainers` based on whether the pod is deleted.

I quite don't understand what the PR is doing now.

Random-Liu replied:
s/sandboes/sandboxes

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -295,15 +313,15 @@ func (cgc *containerGC) evictSandboxes() error {
 	for podUID, sandboxes := range evictSandboxes {
-		if cgc.isPodDeleted(podUID) {
-			// Remove all sandboxes if pod has been deleted.
+		if cgc.isPodDeleted(podUID) || len(activeSandboxes[podUID]) > 0 {
```
1) Why remove all sandboxes when there is active sandbox?
2) Are you sure `sandboxes` is "evictable sandboxes"?

feiskyer replied:
Sorry, this logic is actually wrong.

feiskyer commented:
> For container garbage collection, what we are doing:
> 
> * get evictableContainers (prune out running ones);
> * remove all or keep latest evictableContainers based on whether the pod is deleted.

Yep, but sandboxes are slightly different: 

* sandboxes should not be removed if there are still containers belonging to it. 

Added a new commit, which changed the logic to:

* Get all sandboxes by podUID first, mark sandboxes as active which is ready or with containers
* Sort sandboxes by creation time
* Remove inactive sandboxes
    * remove all inactive sandboxes if pod has been removed
   * else, keep the latest one (regardless of its state)

@Random-Liu PTAL. Feel free to file a new PR if you have different opinions.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -130,12 +132,15 @@ func (cgc *containerGC) removeOldestN(containers []containerGCInfo, toRemove int
 	for i := numToKeep; i < len(sandboxes); i++ {
-		cgc.removeSandbox(sandboxes[i].id)
+		if !sandboxes[i].active {
+			cgc.removeSandbox(sandboxes[i].id)
```
If there is an active sandbox here, returning `sandboxes[:numToKeep]` doesn't make sense to me.

Although we don't use the return value, but the code may confuse people.

feiskyer replied:
Yep, will remove the return value

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -279,31 +298,26 @@ func (cgc *containerGC) evictSandboxes() error {
+	for podUID, sandboxes := range sandboxesByPod {
 		if cgc.isPodDeleted(podUID) {
-			// Remove all sandboxes if pod has been deleted.
+			// Remove all evictable sandboxes if the pod has been removed.
```
Why not keeping the logic the same with container? I didn't see difference except that sandbox with containers is not evictable.

We could still generate an evictable sandboxes list, and use it here.

feiskyer replied:
> In you comment, you said that "keep the latest one (regardless of its state)", but here I didn't see that.

I think you are reviewing an old commit. There is no such comment in latest commit.

feiskyer replied:
> In you comment, you said that "keep the latest one (regardless of its state)", but here I didn't see that.

Sorry, it is in the previous comment. I mean the latest one should be kept, whether it is ready or not. And other active ones are also kept. Will make an update to the comment:

* remove all inactive sandboxes if pod has been removed
* else
  * keep active sandboxes
  * keep the latest sandbox whether it is active or not
  * remove other inactive sandboxes

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -128,6 +132,18 @@ func (cgc *containerGC) removeOldestN(containers []containerGCInfo, toRemove int
+// removeOldestNSandboxes removes the oldest inactive toRemove sandboxes and
+// returns the resulting slice.
+func (cgc *containerGC) removeOldestNSandboxes(sandboxes []sandboxGCInfo, toRemove int) {
+	// Remove from oldest to newest (last to first).
```
Is this `oldest to newest`? Seems to be newest to oldest...

And also the same question for the container remove function.

feiskyer replied:
ACK. Would fix this.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -252,38 +272,48 @@ func (cgc *containerGC) evictSandboxes(minAge time.Duration) error {
+			cgc.removeOldestNSandboxes(sandboxes, len(sandboxes))
+		} else {
+			// Keep latest one if the pod still exists.
+			cgc.removeOldestNSandboxes(sandboxes, len(sandboxes)-1)
```
Maybe explain that we don't need to keep the latest dead sandbox if there is an active one? Since this is different from container gc behavior.

feiskyer replied:
ACK

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -34,6 +34,19 @@ func TestSandboxGC(t *testing.T) {
+	fakePodGetter := m.containerGC.podGetter.(*fakePodGetter)
+	makeGCSandbox := func(pod *v1.Pod, attempt uint32, state runtimeapi.PodSandboxState, withPodGetter bool) sandboxTemplate {
+		if withPodGetter {
+			// initialize the pod getter, explicitly exclude deleted pod
```
Update the comment.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -52,50 +68,51 @@ func TestSandboxGC(t *testing.T) {
 			description: "sandbox with containers should not be garbage collected.",
 			sandboxes: []sandboxTemplate{
-				{pod: pods[0], state: runtimeapi.PodSandboxState_SANDBOX_NOTREADY},
+				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true),
```
s/true/false

Or else sandbox will not be removed because there is corresponding pod.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -52,50 +68,51 @@ func TestSandboxGC(t *testing.T) {
 			sandboxes: []sandboxTemplate{
 				// running sandbox.
-				{pod: pods[0], attempt: 1, state: runtimeapi.PodSandboxState_SANDBOX_READY},
+				makeGCSandbox(pods[0], 1, runtimeapi.PodSandboxState_SANDBOX_READY, true),
```
Should have another exit `pods[0]` to make sure we don't keep latest exit if there is a running one.

feiskyer replied:
ACK.

feiskyer committed:
Update test for sandbox gc

feiskyer commented:
@Random-Liu Updated. PTAL.

Random-Liu commented:
/lgtm
/approve

feiskyer commented:
> @feiskyer Please keep an eye on the test grid to make sure this doesn't introduce flake. :)

Sure.
